stdlib/universe/pivot.go

```diff
@@ -382,3 +390,285 @@ func (t *pivotTransformation) Finish(id execute.DatasetID, err error) {
+	case flux.TFloat:
+		return cr.Floats(j)
+	case flux.TString:
+		return cr.Strings(j)
+	case flux.TBool:
+		return cr.Bools(j)
+	case flux.TTime:
+		return cr.Times(j)
+	default:
+		panic(fmt.Sprintf("unexpected column type: %s", col.Type))
+	}
+}
+
+func (t *pivotTransformation2) UpdateWatermark(id execute.DatasetID, mark execute.Time) error {
+	return t.d.UpdateWatermark(mark)
+}
+
+func (t *pivotTransformation2) UpdateProcessingTime(id execute.DatasetID, mark execute.Time) error {
+	return t.d.UpdateProcessingTime(mark)
+}
+
+func (t *pivotTransformation2) Finish(id execute.DatasetID, err error) {
+	t.groups.Range(func(key flux.GroupKey, value interface{}) {
+		if err != nil {
+			return
+		}
+
+		var tbl flux.Table
+		gr := value.(*pivotTableGroup)
+		tbl, err = gr.doPivot(key, t.alloc)
+		if err != nil {
+			return
+		}
+		err = t.d.Process(tbl)
+	})
+	t.groups.Clear()
+
+	// Inform the downstream dataset that we are finished.
+	t.d.Finish(err)
+}
+
+type pivotTableBuffer struct {
+	keys      []array.Interface
+	valueType flux.ColType
+	values    []array.Interface
+}
+
+func (b *pivotTableBuffer) Insert(k, v array.Interface) {
+	k.Retain()
+	b.keys = append(b.keys, k)
+	v.Retain()
+	b.values = append(b.values, v)
+}
+
+func (b *pivotTableBuffer) Release() {
+	for _, k := range b.keys {
+		k.Release()
+	}
+	for _, v := range b.values {
+		v.Release()
+	}
+}
+
+type pivotTableGroup struct {
+	rowCol  flux.ColMeta
+	buffers map[string]*pivotTableBuffer
+}
+
+func (gr *pivotTableGroup) doPivot(key flux.GroupKey, mem arrowmemory.Allocator) (flux.Table, error) {
+	// Merge all of the keys from each buffer.
+	keys := gr.mergeKeys(mem)
+
+	// Create the table buffer that will be used for the final table.
+	ncols := len(key.Cols()) + len(gr.buffers)
+	tb := &arrow.TableBuffer{
+		GroupKey: key,
+		Columns:  make([]flux.ColMeta, 0, ncols),
+		Values:   make([]array.Interface, 0, ncols),
+	}
+	tb.Columns = append(tb.Columns, gr.rowCol)
+	tb.Values = append(tb.Values, keys)
+
+	// Add the group key columns to the table.
+	for j, col := range key.Cols() {
+		tb.Columns = append(tb.Columns, col)
+		tb.Values = append(tb.Values, arrow.Repeat(key.Value(j), keys.Len(), mem))
+	}
+
+	// Build each column by appending a value when it matches
+	// with one of the keys and null when it does not.
+	labels := make([]string, 0, len(gr.buffers))
+	for label := range gr.buffers {
+		labels = append(labels, label)
+	}
+	sort.Strings(labels)
+
+	for _, label := range labels {
+		buf := gr.buffers[label]
+		vs := gr.buildColumn(keys, buf, mem)
+		tb.Columns = append(tb.Columns, flux.ColMeta{
+			Label: label,
+			Type:  buf.valueType,
+		})
+		tb.Values = append(tb.Values, vs)
+		buf.Release()
+	}
+
+	if err := tb.Validate(); err != nil {
+		return nil, err
+	}
+	return tableutil.FromBuffer(tb), nil
+}
+
+//go:generate -command tmpl ../../gotool.sh github.com/benbjohnson/tmpl
+//go:generate tmpl -data=@../../internal/types.tmpldata -o pivot.gen.go pivot.gen.go.tmpl
```

Comment:
nit, can you move these to the top of the file?

Reply:
Sure. I put it at the bottom because that's where the code originally was, but it makes sense at the top too.